lib/plugins/package/lib/zipService.js

```diff
@@ -1,73 +1,188 @@
 'use strict';
 
+const which = require('which');
 const archiver = require('archiver');
+const exec = require('child_process').exec;
 const BbPromise = require('bluebird');
 const path = require('path');
 const fs = require('fs');
 const glob = require('glob-all');
 
-module.exports = {
-  zipDirectory(servicePath, exclude, include, zipFileName) {
-    const patterns = ['**'];
-
-    exclude.forEach((pattern) => {
-      if (pattern.charAt(0) !== '!') {
-        patterns.push(`!${pattern}`);
-      } else {
-        patterns.push(pattern.substring(1));
-      }
-    });
+/**
+ * Combines exclude and include patterns into one array
+ * respecting pattern overrides
+ * @param {String[]} exclude Array of exclude patterns
+ * @param {String[]} include Array of include patterns
+ * @return {String[]} Combined array of patterns
+ */
+function getFilesPatterns(exclude, include) {
+  const patterns = ['**'];
 
-    // push the include globs to the end of the array
-    // (files and folders will be re-added again even if they were excluded beforehand)
-    include.forEach((pattern) => {
-      patterns.push(pattern);
-    });
+  exclude.forEach((pattern) => {
+    if (pattern.charAt(0) === '!') {
+      patterns.push(pattern.substring(1));
+    } else {
+      patterns.push(`!${pattern}`);
+    }
+  });
 
-    const zip = archiver.create('zip');
+  // push the include globs to the end of the array
+  // (files and folders will be re-added again even if they were excluded beforehand)
+  include.forEach((pattern) => {
+    patterns.push(pattern);
+  });
 
-    const artifactFilePath = path.join(
-      servicePath,
-      '.serverless',
-      zipFileName
-    );
+  return patterns;
+}
 
-    this.serverless.utils.writeFileDir(artifactFilePath);
+/**
+ * Finds files matching given patterns and excluding directories
+ * @param {String} servicePath Service path for searching
+ * @param {String[]} patterns Array of globs to use for file matching
+ */
+function getFilesToZip(servicePath, patterns) {
+  return BbPromise.fromCallback((cb) => {
+    // sync version returns temp zip files from .serverless dir
+    glob(patterns, {
+      cwd: servicePath,
+      dot: true,
+      silent: true,
+      follow: true,
+    }, cb);
+  }).filter((filePath) => {
+    const stats = fs.statSync(path.resolve(servicePath, filePath));
+    return !stats.isDirectory(filePath);
+  });
+}
 
-    const output = fs.createWriteStream(artifactFilePath);
+/**
+ * Checks whether machine has zip binary in PATH
+ * return {Promise.<Boolean>} Is available - true/false
+ */
+function isZipAvailable() {
+  return new BbPromise((resolve) => {
+    which('zip', (error) => resolve(!error));
+  });
+}
 
-    output.on('open', () => {
-      zip.pipe(output);
+/**
+ * Get a list of arguments for zip command each less than ~100k chars
```

Comment:
@ryansb thanks for the clarification.
Seems that I was misreading it. I (somehow) thought that the argument is the content of the file.

Splitting it up to mitigate problems with ver large services makes sense 👍